### PR TITLE
Add optional Paid Status on Order

### PR DESCRIPTION
**Pull request by frederic34**
By example, when I import orders from prestashop, I want to be able to see if they are paid or not.

### CODE REVIEW

**Comment by marcosgdf** [comment]
A customer order is not supposed to have a paid/not paid status... An invoice should be created from a customer order and that is the one that should be marked as paid

**Comment by frederic34** [comment]
When I have an order in prestashop or in marketplace, I will not send it until the order is paid, but I create the order to reserve  product until it is paid. If order is not paid, it is canceled and product is no more reserved.

**Comment by marcosgdf** [comment]
I understand... but I am looking it from the accountant view and only invoices have the "right" to be paid…

**Comment by frederic34** [comment]
It's optional and I'm not creating payments, it's just a flag.

**Comment by eldy** [comment]
You we must not have on trigger per field, but one trigger per action (CREATE / MODIFY / DELETE). So when you set the status, you must call the existing trigger MODIFY (you can provide info to trigger by setting the $object->context property if you need) and removed addition of trigger SET_PAID.

NOTE 2:
i don't understand the change. There is already a step "Set paid" for customer order. Stored into field "facture"
 See more info on page
http://wiki.dolibarr.org/index.php/Commande

**Comment by frederic34** [comment]
I think order paid and order billed are different

``` php
Draft --> Validated --> Paid --> Sending --> Delivered --------> Completed (=Delivered + Billed)
                      \         \             \                             /
                       \-----------------> Invoice ------------------------/

Customer in account
Draft --> Validated --> Paid --> Sending --> Delivered ---------------------> Completed (=Delivered + Billed)
Draft --> Validated --> Paid --> Sending --> Delivered ---------------------> Completed (=Delivered + Billed)
                                              \                                                /
                                               \ - Invoice at end of month for several orders-/

Draft     : do not decrease available stock
Validated : we decrease available stock to sell
Paid      : ok to send
Sending   : we decrease real stock

```

**Comment by BebZ** [comment]
I think the same way of Frederic, 
sometimes there's a delay between registering the payment on bank account and checking if order is paid.
In many companies it works like this : 
If an order with no invoice attached is not paid (first time order from prospect) 

> the order is not sent, 
> or it has to be a real "trustful" prospect (but there is no way to flag it as "trustful")... so "paid" order is a good choice.

Sometimes the prospect sends you the proof of payment by bank wire transfer, but you receive it on your bank account the day after, so you record payment on invoice only 1 day after 
= 1 day lost on the sending.
A "paid" order can be sent, without an invoice paid linked... this is the "trick" ?

So maybe the word "paid" is source of misunderstanding, maybe "checked" or "shippable" or anythig else... Poll, suggestions ?

**Comment by eldy** [comment]
@frederic34 I don't understand your explanation saying order payed != order billed. Of course it is same, it is a synonim. But i understand explanation of @BebZ .
Was you thinking of  BebZ case ?  In your example if invoice is generated at end of month after shipment is done, we can't called the intermediate status "paid", so we should called it "Ready for shipment" or "shippable" instead, shouldn't it ?

**Comment by BebZ** [comment]
@frederic34 you said : It's optional and I'm not creating payments, it's just a flag.
so it doesn't matter whatever it is called ? 
it's just a way for you to know if the customer has "paid" on the website 
so you can send it, then generate an invoice (which, this one will be "paid"), one or 2 days after when you will have received your bank receipt with the payment :)

For me it was the same problem (paid on website, but not possible to make an invoice and emit payment before the money is physically on bank account), 

this intermeditate status will solve it. Thanks !

**Comment by BebZ** [comment]
Ping @frederic34 ?
